### ISSUE TITLE
Incorrect type inference with a class generic over type passed in `__init__`.

### ISSUE THREAD

**Issue by elenakrittik** [microsoft/pyright#6893]
**Describe the bug**

When trying to instantiate a class generic over V whose `__init__` accepts an argument `type_: type[V]` with `type_=Literal["string literal"]`, V gets inferred to the more general `str` instead of expected `Literal["string literal"]`.

**Code or Screenshots**

```python
import typing as t

V = t.TypeVar("V")

class Key(t.Generic[V]):
    def __init__(self, type_: t.Type[V]) -> None:
        ...

my_key = Key(t.Literal["hi"])
reveal_type(my_key) # Type of "my_key" is "Key[str]"
                    # expected "Type of "my_key" is "Key[Literal['hi']]""
```

**VS Code extension or command-line**
pyright 1.1.342, VSCode extension behaves the same but i unfortunately don't have exact version at hand right now.

**Comment by erictraut** [comment]
Pyright is working correctly here, so I don't consider this a bug, at least not as it was reported.

You are using the `typing.Literal` special form in an invalid manner in your code. It should appear only within a type annotation, but you're passing it as a runtime value to the `Key` constructor.

Pyright should arguably generate an error if one attempts to pass `Literal` to a parameter annotated as `type[T]` since `Literal` is not a class. I've opened [a new issue](https://github.com/microsoft/pyright/issues/6895) to track this.

I'm not sure what you're trying to do with this code, but perhaps this is what you intended?

```python
from typing import Generic, LiteralString, TypeVar

V = TypeVar("V", bound=LiteralString)

class Key(Generic[V]):
    def __init__(self, value: V) -> None:
        ...

my_key = Key("hi")
reveal_type(my_key)
```
